### PR TITLE
GH-39815: [C++] Document and micro-optimize ChunkResolver::Resolve()

### LINKED ISSUE #39815: [C++] Document the guarantees of ChunkResolver::Resolve and minimize branching
### Describe the enhancement requested

The invariants of `ChunkResolver` and pre-conditions of `ChunkResolver::Resolve()` are unclear which can lead to multiple problems:
1) Misuse
2) Misunderstanding that leads to addition of more checks that can kill speculative execution in tight loops
3) Missed micro-optmization opportunities

Request: document the class and implement possible micro-optmizations.

### Component(s)

C++

### PR DESCRIPTION
**Pull request by felipecrv**
### Rationale for this change

There has been interest in improving operations on chunked-arrays and even though `ChunkResolver::Resolve()` is not a big contributor in most kernels, the fact that it can be used from tight loops warrants careful attention to branch prediction and memory effects of its implementation.

### What changes are included in this PR?

 - Documentation of invariants and behavior of functions
 - Multiple optimizations justified by microbenchmarks
 - Addition of a variation of `Resolve` that takes a hint as parameter
 - Fix of an out-of-bounds memory access that doesn't affect correctness (it can only reduce effectiveness of cache in very rare situations, but is nevertheless an issue)

### Are these changes tested?

Yes, by existing tests.

### Are there any user-facing changes?

 - The `arrow::internal::ChunkResolver::Bisect()` function was `protected` and is now `private` with a different signature
* Closes: #39815

### CODE REVIEW

**Comment by felipecrv** [comment]
Benchmarks of `sort` and `rank` on chunked arrays -- heavy users of `ChunkResolver`. 3 measurements after roughly every change to give an idea of level of noise. The purple group (`bounds-check-fix`) is when I fixed the out-of-bounds access bug that exists on `main` (not introduced by me in the optimizations). After that, the other two groups bring improvements that bring the throughput back to what was achieved before the bounds check.

resolve

Ideas that were tried and didn't make a difference or made throughput worse:
 - Removing the use of `std::atomic` completely, relaxed atomic operations are enough (which is good because that could introduce bugs)
 - Starting the `Bisect` on different ranges depending on the results of the branches
 

[1] `ninja arrow-compute-vector-sort-benchmark && ./**/arrow-compute-vector-sort-benchmark --benchmark_filter="ChunkedArray(Sort|Rank).*Int64.*65536/100(/tiebreaker:2|$)" --benchmark_out_format=csv`

**Comment by felipecrv** [comment]
@amol- @pitrou @js8544

**Comment by pitrou** [comment]
+1 on the principle, looks like a nice improvement.

**Comment by pitrou** [comment]
@ursabot please benchmark

File: cpp/src/arrow/compute/kernels/vector_sort.cc [hunk 1]
```diff
@@ -748,11 +749,15 @@ class TableSorter {
...
 
+    ChunkLocation left_loc{0, 0};
+    ChunkLocation right_loc{0, 0};
     std::merge(nulls_begin, nulls_middle, nulls_middle, nulls_end, temp_indices,
                [&](uint64_t left, uint64_t right) {
                  // First column is either null or nan
-                 const auto left_loc = left_resolver_.Resolve(left);
-                 const auto right_loc = right_resolver_.Resolve(right);
+                 left_loc =
+                     left_resolver_.ResolveWithChunkIndexHint(left, left_loc.chunk_index);
```
Comment (pitrou) [inline cpp/src/arrow/compute/kernels/vector_sort.cc:758]: The point of this is to avoid all atomic accesses, right?

Comment (felipecrv) [reply cpp/src/arrow/compute/kernels/vector_sort.cc:758]: Memory access really: keeping the `ChunkLocation::chunk_index` in registers throughout the loop -- of course it depends on the register allocation and inlining of `ResolveWithChunkIndexHint`. It improved the sort benchmarks slightly.

**Review by pitrou — APPROVED** [review]
There are nice chunked Sort and Rank performance improvements on the ARM64 benchmark machines. Thank you!

**Pull request merged**